lib/search_files.js

```diff
@@ -0,0 +1,291 @@
+        Array.prototype.push.apply(ret, v);
+      } else {
+        ret.push(v);
+      }
+    }
+  }
+  return ret;
+}
+
+function eachProp(obj, fn /*, ...*/) {
+  var args = Array.prototype.slice.call(arguments, 2);
+
+  var ret = {};
+  var keys = Object.keys(obj);
+  for (var i = 0, n = keys.length; i < n; i++) {
+    var key = keys[i];
+    ret[key] = fn.apply(obj, [obj[key], key].concat(args));
+  }
+  return ret;
+}
+
+
+function parsePath(pth) {
+  return path.parse ? path.parse(pth) : require('path-parse')(pth);
+}
+
+function isAbsolutePath(pth) {
+  return path.isAbsolute ? path.isAbsolute(pth) : !!parsePath(pth).root;
+}
+
+function normalizeHomedir(pth, homedir) {
+  if (!pth) {
+    return pth;
+  }
+
+  if (pth[0] === '~') {
+    if (pth.length === 1) {
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
+      return path.join(homedir, pth.slice(2));
+    }
+
+    if (isWindows() && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
+    }
+  }
+
+  return pth;
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
+    pth = pth();
+  }
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
+    return path.join(ctx.cwd, pth);
+  }
+
+  return path.normalize(pth);
+}
+
+
+var initCtx = {
+  homedir: os.homedir ? os.homedir() : require('userhome')(),
+  cwd: process.cwd(),
+};
+
+function normalizeContext(ctx) {
+  ctx = eachProp(ctx || {}, doNothing);
+
+  ctx.cwd = !ctx.cwd ? initCtx.cwd :
+    normalizePath(ctx.cwd, initCtx);
+
+  ctx.homedir = !ctx.homedir ? initCtx.homedir :
+    normalizePath(ctx.homedir, initCtx);
+
+  return ctx;
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
+    exts = exts();
+  }
+
+  if (isString(exts)) {
+    return [exts];
+  }
+
+  if (isArray(exts)) {
+    return exts.filter(isString);
+  }
+
+  if (isObject(exts)) {
+    return Object.keys(exts).sort();
+  }
+
+  return [''];
+}
+
+function findFile(pth, basedir, findUp) {
+
+  var parsed = parsePath(pth);
+  if (parsed.root) { // pth is absolute path
+    findUp = false;
+    basedir = parsed.root;
+    pth = pth.slice(parsed.root.length);
+  }
+
+  var globOpt = { nocase: true, cwd: basedir };
+
+  if (isWindows()) {
+    pth = pth.replace(/\\/g, '/');
+  }
+
+  if (!findUp) {
+    return eachUnNullElem(glob.sync(pth, globOpt), toFullPath, globOpt.cwd);
+  }
+
+  var lastPath;
+  do {
+    var founds = glob.sync(pth, globOpt);
+    if (founds.length > 0) {
+      return eachUnNullElem(founds, toFullPath, globOpt.cwd);
+    }
+
+    lastPath = globOpt.cwd;
+    globOpt.cwd = path.resolve(globOpt.cwd, '..');
+
+  } while (globOpt.cwd !== lastPath);
+
+  return [];
+
+  function toFullPath(pth, basedir) {
+    return path.join(basedir, pth);
+  }
+}
+
+function checkSymbolicLink(pth) {
+  try {
+    if (!fs.lstatSync(pth).isSymbolicLink()) {
+      return pth;
+    }
+
+    if (fs.existsSync(fs.realpathSync(pth))) {
+      return pth;
+    }
+
+  } catch (e) {}
```

Comment:
Ah, since this `pth` file always exists, this `lstatSync` does not throw Error. I'll remove this try-catch.